Log messages are lost when an exception is thrown

Hello

**Describe the bug**
Im not quite sure if this is a bug of the java instrumentation or something else but I hope you can guide me somehow towards a solution.

We recently switched from Xray SDK to AWS ADOT via javaagent and sidecar collector. In general this works pretty good and a lot better than Xray itself but we noticed some issues especially when an exception is thrown. 
We have a small Spring Boot app that uses standard logback logging via [logstash-logback-encoder](https://github.com/logstash/logstash-logback-encoder).  It happens in special cases that logs are completely dropped when an exception is thrown deep in the call hierarchy. It doesn't happen for all exceptions, it seems to be related to the depth of call/trace. What I mean by log messages are dropped is that we cant see any error handling log output from the enclosing classes. It just seems to be stuck at some point. 

What I tried so far:
- disabled the logstash encoder stuff, basically switched back to standard out logging via logback -> same behaviour
- disabled javaagent -> logs come back
- adding a try/catch block somewhere in the middle around the problematic code -> logs come back
- set the agent to debug log -> no errors or warning around this time, seems quite normal

**Steps to reproduce**
To be honest I don't know how to produce this if its really related to the call depth. I can try if you want me to do it :)

**What did you expect to see?**
Logs are still shipped/printed.

**What did you see instead?**
No logs.

**What version are you using?**
v0.15.0 (re-packaged version for AWS https://github.com/aws-observability/aws-otel-java-instrumentation)
Spring Boot 2.4.2


**Environment**
Compiler: corretto 1.8
OS: debian

I know this is somehow a bad bug report, but I don't know where to look at.

Thank you very much.

I was able to setup a reproduction repo:
https://github.com/otbe/factcast/tree/adot-issue
 
After you build the project by running `mvn install -DskipTests` you should able to run the test `org.factcast.itests.factus.FactusClientTest.testSubscription`. The test will fail 
and log something like

```
Unhandled onError:
io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.
...
Caused by: java.lang.IllegalStateException: broken
```

If you now add the javaagent (pure opentelemtry or aws flavoured one) to the test execution the test will just hang. So its more problematic than I thought because its not about logs getting dropped its more about the execution seems to hang for some reason. If you comment out the `IllegalStateException` [here](https://github.com/otbe/factcast/blob/adot-issue/factcast-itests/factcast-itests-factus/src/test/java/org/factcast/itests/factus/proj/SubscribedUserNames.java#L43) the test fails (due to assertion errors).

--- a/instrumentation/grpc-1.5/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_5/client/TracingClientInterceptor.java
+++ b/instrumentation/grpc-1.5/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_5/client/TracingClientInterceptor.java
@@ -142,6 +142,7 @@ public class TracingClientInterceptor implements ClientInterceptor {
         delegate().onMessage(message);
       } catch (Throwable e) {
         tracer.addThrowable(span, e);
+        throw e;
       }
     }
 

--- a/instrumentation/grpc-1.5/testing/src/main/groovy/io/opentelemetry/instrumentation/grpc/v1_5/AbstractGrpcTest.groovy
+++ b/instrumentation/grpc-1.5/testing/src/main/groovy/io/opentelemetry/instrumentation/grpc/v1_5/AbstractGrpcTest.groovy
@@ -158,7 +158,9 @@ abstract class AbstractGrpcTest extends InstrumentationSpecification {
     client.sayHello(Helloworld.Request.newBuilder().setName(paramName).build())
 
     then:
-    thrown StatusRuntimeException
+    def e = thrown(StatusRuntimeException)
+    e.status.code == grpcStatus.code
+    e.status.description == grpcStatus.description
 
     assertTraces(1) {
       trace(0, 2) {
@@ -242,7 +244,10 @@ abstract class AbstractGrpcTest extends InstrumentationSpecification {
     client.sayHello(Helloworld.Request.newBuilder().setName(paramName).build())
 
     then:
-    thrown StatusRuntimeException
+    def e = thrown(StatusRuntimeException)
+    // gRPC doesn't appear to propagate server exceptions that are thrown, not onError.
+    e.status.code == Status.UNKNOWN.code
+    e.status.description == null
 
     assertTraces(1) {
       trace(0, 2) {
@@ -444,6 +449,67 @@ abstract class AbstractGrpcTest extends InstrumentationSpecification {
         }
       }
     }
+
+    cleanup:
+    channel?.shutdownNow()?.awaitTermination(10, TimeUnit.SECONDS)
+    server?.shutdownNow()?.awaitTermination()
+  }
+
+  // Regression test for https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2285
+  def "client error thrown"() {
+    setup:
+    BindableService greeter = new GreeterGrpc.GreeterImplBase() {
+      @Override
+      void sayHello(
+        final Helloworld.Request req, final StreamObserver<Helloworld.Response> responseObserver) {
+        // Send a response but don't complete so client can fail itself
+        responseObserver.onNext(Helloworld.Response.getDefaultInstance())
+      }
+    }
+    def port = PortUtils.randomOpenPort()
+    Server server
+    server = configureServer(ServerBuilder.forPort(port)
+      .addService(greeter))
+      .build().start()
+    ManagedChannelBuilder channelBuilder
+    channelBuilder = configureClient(ManagedChannelBuilder.forAddress("localhost", port))
+
+    // Depending on the version of gRPC usePlainText may or may not take an argument.
+    try {
+      channelBuilder.usePlaintext()
+    } catch (MissingMethodException e) {
+      channelBuilder.usePlaintext(true)
+    }
+    ManagedChannel channel = channelBuilder.build()
+    def client = GreeterGrpc.newStub(channel)
+
+    when:
+    AtomicReference<Throwable> error = new AtomicReference<>()
+    CountDownLatch latch = new CountDownLatch(1)
+    client.sayHello(
+      Helloworld.Request.newBuilder().setName("test").build(),
+      new StreamObserver<Helloworld.Response>() {
+        @Override
+        void onNext(Helloworld.Response r) {
+          throw new IllegalStateException("illegal")
+        }
+
+        @Override
+        void onError(Throwable throwable) {
+          error.set(throwable)
+          latch.countDown()
+        }
+
+        @Override
+        void onCompleted() {
+          latch.countDown()
+        }
+      })
+
+    latch.await(10, TimeUnit.SECONDS)
+
+    then:
+    error.get() != null
 
     cleanup:
     channel?.shutdownNow()?.awaitTermination(10, TimeUnit.SECONDS)

